### DISCUSSION TITLE
SeeedStudio SenseCAP Indicator - SPI question

### DISCUSSION THREAD

**Discussion by u4mzu4** [moononournation/Arduino_GFX#334] [General]
Hello, and first of all thanks for this nice library!
I have a new gadet from SeeedStudio, a [SenseCAP Indicator](https://www.seeedstudio.com/SenseCAP-Indicator-D1S-p-5645.html). It's an ESP32-S3 based 4" display, looks like a simple [ESP32-S3-RGB](https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration#esp32-s3-rgb).
However!
For some mysterious reason the SPI_CS and the LCD_RST pins were outsourced to a PCA9535 I2C expander, you can find a block diagram [HERE.](https://files.seeedstudio.com/wiki/SenseCAP/SenseCAP_Indicator/SenseCAP_Indicator_6.png)

I've seen that CS can be GFX_NOT_DEFINED (set fix LOW), so my setting are the following:
```
#define GFX_BL 45
Arduino_DataBus *bus = new Arduino_SWSPI(
    GFX_NOT_DEFINED /* DC */, GFX_NOT_DEFINED /* CS */,
    41 /* SCK */, 48 /* MOSI */, GFX_NOT_DEFINED /* MISO */);

// option 1:
// Uncomment for 4" rect display
Arduino_ESP32RGBPanel *rgbpanel = new Arduino_ESP32RGBPanel(
    18 /* DE */, 17 /* VSYNC */, 16 /* HSYNC */, 21 /* PCLK */,
    4 /* R0 */, 3 /* R1 */, 2 /* R2 */, 1 /* R3 */, 0 /* R4 */,
    10 /* G0 */, 9 /* G1 */, 8 /* G2 */, 7 /* G3 */, 6 /* G4 */, 5 /* G5 */,
    15 /* B0 */, 14 /* B1 */, 13 /* B2 */, 12 /* B3 */, 11 /* B4 */,
    1 /* hsync_polarity */, 10 /* hsync_front_porch */, 8 /* hsync_pulse_width */, 50 /* hsync_back_porch */,
    1 /* vsync_polarity */, 10 /* vsync_front_porch */, 8 /* vsync_pulse_width */, 20 /* vsync_back_porch */);
Arduino_RGB_Display *gfx = new Arduino_RGB_Display(
    480 /* width */, 480 /* height */, rgbpanel, 0 /* rotation */, true /* auto_flush */,
    bus, GFX_NOT_DEFINED /* RST */, st7701_type1_init_operations, sizeof(st7701_type1_init_operations));
```
Compiled PDQgraphicstest and running, the Serial output is the following:
    
Benchmark	micro-secs
Screen fill	85303
Text	28389
Pixels	477361
Lines	598158
Horiz/Vert Lines	81183
Rectangles (filled)	496544
Rectangles (outline)	67436
Triangles (filled)	209196
Triangles (outline)	45312
Circles (filled)	48793
Circles (outline)	57702
Arcs (filled)	48146
Arcs (outline)	151570
Rounded rects (filled)	535074
Rounded rects (outline)	73095
Done!

Blacklight is OK, but I can not see anything on the display. Looks like the init was wrong, is there any solution to bitbang the CS (eg. though some macro), or did I screw something else up?

**Comment by u4mzu4** [comment]
Ooh-ahh, it's working!
I wrote (messed up) my own Arduino_DataBus, which is a mixture of Arduino_SWSPI and Arduino_XL9535SWSPI (for CS). Ugly source code, full of inelegant solutions, but it's working.
@moononournation: if you need "official" support for this device, we can clean/merge the source code, anyway I can go further with my own workaround.

**Reply by mc32768** [reply]
Would you be willing to share the code workaround that made the screen functional for you?  I am stuck at the same spot where you were previously (Serial says everything works, backlight is on, but display remains blank).  I've tried modifying Arduino_SWSPI so that CS_HIGH and CS_LOW use the PCA9535, but that hasn't made it work - I feel like I'm missing a step somewhere.  Thanks! :)

**Reply by u4mzu4** [reply]
Hello!
I just used the driver [https://github.com/semcneil/PCA95x5](https://github.com/semcneil/PCA95x5), but the "builtin" XL9535 would be better I think.
Here is my working demo (just rename to .cpp):
[Arduino_SWSPI.txt](https://github.com/moononournation/Arduino_GFX/files/12571881/Arduino_SWSPI.txt)

The SPI speed is not critical because only the init happens there.
If you want to rotate the display you can use the X-direction control command below as well.

**Reply by mc32768** [reply]
Thanks so much!  It turns out I had done almost _exactly_ the same thing as you, even using the same PCA95x5 library.  The mistake was that I forgot to specify the pins in Wire.begin(39, 40, 400000) so that is why the PCA9535 wasn't working for me.  It was really helpful to be able to check out your code and I appreciate your willingness to share it.  

One minor difference I noticed was that the Seeed ESP-IDF sample code (components\bsp\src\boards\sensecap_indicator_board.c) seems to leave the PCA9535 pins in their default state (inputs) and then it specifically sets only 4,5,7,8 and 10 as outputs.  Whereas in the demo code linked above, it is setting all pins to outputs (ioex.direction(PCA95x5::Direction::OUT_ALL)).  Seeed's block diagram does show a one-way arrow coming from the PCA9535, implying all pins are outputs, but when I looked at the touch panel datasheet it says the INT pin for the touch panel (IO0_6 on PCA9535) is an _output_ of the touch panel, so if it's also set as an output of the PCA9535, that could cause a conflict.  It might be safer to set only specific known pins as outputs and leave the rest as inputs.

I am glad to finally see the Sensecap Indicator working - the device is new, so it's hard to find documentation or functional sample code at the moment, and the design seems to have quite a few quirks.  Thanks again for taking the time to help me with this!

**Reply by u4mzu4** [reply]
@mc32768 
I'm glad that you was able to use my trashy workaround. That would be great if we can clenaup and add as official solution. Anyway, regardinig your issue with the directions: if you check the [block diagram](https://files.seeedstudio.com/wiki/SenseCAP/SenseCAP_Indicator/SenseCAP_Indicator_6.png) carefully, you can find the GPIO45 as interrupt of the touch panel, which is really nice (instead of polling in every n millis). However you can find this GPIO as above
`#define GFX_BL 45`
so, this is a clear output for backlight (even in the Seeed code base), and never used (unforunately) as input.
So summary: there is **no input for PCA9535** unfortunately.

**Reply by mc32768** [reply]
I wonder if there may be a mistake in that block diagram, since they have ESP32 GPIO45 shown in two places.  It is shown as PWM backlight control (output of ESP32) and as PCA9535 interrupt (input of ESP32, output of PCA9535).  It does not seem logical that those two functions would coexist on the same pin.  I agree with you that it appears to be used for backlight.

In my earlier message, the input signal I was referring to is between the PCA9535 and the touch panel controller (TP_INT on IO0_6 of the PCA9535).  I think that one should be configured as an input to the PCA9535 since it is definitely an output of the touch panel controller.  It would be bad if both devices (PCA9535 and touch panel controller) were driving it concurrently so that's why I was thinking it should be configured as an input on PCA9535, even if it's never used in code anywhere else.  

I haven't actually gotten to the point of making the touch panel work yet so I might know more if I can proceed to the point where I need to use it.

**Reply by u4mzu4** [reply]
You're right, IO0_6 is an input on the PCA9535, I've used as input as well when I played with the touch driver, and the elegant configuration is the input. My pain is that it can't be used for anything.

**Comment by u4mzu4** [comment]
In the meantime I've just run into a new issue: the _st7701_type1_init_operations_ is upside down on this device, and the SW rotation (2 /* rotation */) skips a few lines at the bottom of the display in case of big/full frame buffer.
I've checked the original SeeedStudio init sequence and found the line:
`WRITE_C8_D8, 0xC7, 0x04,`
which is a 
SDIR (C7h/C700): X-direction Control 
based on the ST7701S datasheet.
So maybe you can use this above as a 180° HW rotation solution.

**Reply by chalei** [reply]
`WRITE_C8_D8, 0xC7, 0x04,`

where should I put this in **Arduino_RGB_Display.h** file

```
static const uint8_t st7701_type1_init_operations[] = {
    BEGIN_WRITE,
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x10,

    WRITE_C8_D16, 0xC0, 0x3B, 0x00,
    WRITE_C8_D16, 0xC1, 0x0D, 0x02,
    WRITE_C8_D16, 0xC2, 0x31, 0x05,
    WRITE_C8_D8, 0xCD, 0x08,

    WRITE_COMMAND_8, 0xB0, // Positive Voltage Gamma Control
    WRITE_BYTES, 16,
    0x00, 0x11, 0x18, 0x0E,
    0x11, 0x06, 0x07, 0x08,
    0x07, 0x22, 0x04, 0x12,
    0x0F, 0xAA, 0x31, 0x18,

    WRITE_COMMAND_8, 0xB1, // Negative Voltage Gamma Control
    WRITE_BYTES, 16,
    0x00, 0x11, 0x19, 0x0E,
    0x12, 0x07, 0x08, 0x08,
    0x08, 0x22, 0x04, 0x11,
    0x11, 0xA9, 0x32, 0x18,

    // PAGE1
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x11,

    WRITE_C8_D8, 0xB0, 0x60, // Vop=4.7375v
    WRITE_C8_D8, 0xB1, 0x32, // VCOM=32
    WRITE_C8_D8, 0xB2, 0x07, // VGH=15v
    WRITE_C8_D8, 0xB3, 0x80,
    WRITE_C8_D8, 0xB5, 0x49, // VGL=-10.17v
    WRITE_C8_D8, 0xB7, 0x85,
    WRITE_C8_D8, 0xB8, 0x21, // AVDD=6.6 & AVCL=-4.6
    WRITE_C8_D8, 0xC1, 0x78,
    WRITE_C8_D8, 0xC2, 0x78,

    WRITE_COMMAND_8, 0xE0,
    WRITE_BYTES, 3, 0x00, 0x1B, 0x02,

    WRITE_COMMAND_8, 0xE1,
    WRITE_BYTES, 11,
    0x08, 0xA0, 0x00, 0x00,
    0x07, 0xA0, 0x00, 0x00,
    0x00, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE2,
    WRITE_BYTES, 12,
    0x11, 0x11, 0x44, 0x44,
    0xED, 0xA0, 0x00, 0x00,
    0xEC, 0xA0, 0x00, 0x00,

    WRITE_COMMAND_8, 0xE3,
    WRITE_BYTES, 4, 0x00, 0x00, 0x11, 0x11,

    WRITE_C8_D16, 0xE4, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE5,
    WRITE_BYTES, 16,
    0x0A, 0xE9, 0xD8, 0xA0,
    0x0C, 0xEB, 0xD8, 0xA0,
    0x0E, 0xED, 0xD8, 0xA0,
    0x10, 0xEF, 0xD8, 0xA0,

    WRITE_COMMAND_8, 0xE6,
    WRITE_BYTES, 4, 0x00, 0x00, 0x11, 0x11,

    WRITE_C8_D16, 0xE7, 0x44, 0x44,

    WRITE_COMMAND_8, 0xE8,
    WRITE_BYTES, 16,
    0x09, 0xE8, 0xD8, 0xA0,
    0x0B, 0xEA, 0xD8, 0xA0,
    0x0D, 0xEC, 0xD8, 0xA0,
    0x0F, 0xEE, 0xD8, 0xA0,

    WRITE_COMMAND_8, 0xEB,
    WRITE_BYTES, 7,
    0x02, 0x00, 0xE4, 0xE4,
    0x88, 0x00, 0x40,

    WRITE_C8_D16, 0xEC, 0x3C, 0x00,

    WRITE_COMMAND_8, 0xED,
    WRITE_BYTES, 16,
    0xAB, 0x89, 0x76, 0x54,
    0x02, 0xFF, 0xFF, 0xFF,
    0xFF, 0xFF, 0xFF, 0x20,
    0x45, 0x67, 0x98, 0xBA,

    //-----------VAP & VAN---------------
    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x13,

    WRITE_C8_D8, 0xE5, 0xE4,

    WRITE_COMMAND_8, 0xFF,
    WRITE_BYTES, 5, 0x77, 0x01, 0x00, 0x00, 0x00,

    WRITE_COMMAND_8, 0x21,   // 0x20 normal, 0x21 IPS
    WRITE_C8_D8, 0x3A, 0x60, // 0x70 RGB888, 0x60 RGB666, 0x50 RGB565

    WRITE_COMMAND_8, 0x11, // Sleep Out
    END_WRITE,

    DELAY, 120,

    BEGIN_WRITE,
    WRITE_COMMAND_8, 0x29, // Display On
    END_WRITE};
```

is it in the beginning of the st7701 definition or the end?

**Reply by u4mzu4** [reply]
@chalei

This is the original Seed solution:
![image](https://github.com/moononournation/Arduino_GFX/assets/39950568/58ac3908-d0d2-4c29-82b5-c04a711dc7ec)
